src/pytest_container/container.py

```diff
@@ -41,6 +41,75 @@ def __str__(self) -> str:
         return "oci" if self == ImageFormat.OCIv1 else "docker"
 
 
+@enum.unique
+class NetworkProtocol(enum.Enum):
+    """Network protocols support for port forwarding."""
+
+    #: Transmission Control Protocol
+    TCP = "tcp"
+    #: User Datagram Protocol
+    UDP = "udp"
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class PortForwarding:
+    container_port: int
+    host_port: int = -1
+    protocol: NetworkProtocol = NetworkProtocol.TCP
+
+    @property
+    def forward_cli_args(self) -> List[str]:
+        return [
+            "-p",
+            f"{self.host_port}:{self.container_port}/{self.protocol}",
+        ]
+
+    def __str__(self) -> str:
+        return str(self.forward_cli_args)
+
+
+def is_socket_listening_on_localhost(
+    port: int, protocol: NetworkProtocol
+) -> bool:
+    """Checks if a socket is listening on localhost for the given port and
+    protocol."""
+
+    open_ports = [
+        int(sock[6:].split(":")[-1])
+        for sock in LOCALHOST.socket.get_listening_sockets()
+        if sock[:3] == str(protocol)
+    ]
+    return port in open_ports
+
+
+def set_port_forwards(
+    port_forwards: List[PortForwarding],
+) -> List[PortForwarding]:
+    START_PORT = 1025
+    MAX_PORT = 2**16
+    finished_forwards: List[PortForwarding] = []
+
+    for port in port_forwards:
+        for i in range(START_PORT, MAX_PORT):
+            if not is_socket_listening_on_localhost(i, port.protocol):
+                finished_forwards.append(
```

Comment:
Could it be useful here the _testinfra_ method `LOCALHOST.socket(f"...:{port}").is_listening`, like in [webserver_1]( https://github.com/SUSE/BCI-tests/blob/df22cdccd29a46e5cb39b69281b55ecfa5c7bdc3/tests/test_python.py#L144) ?

Reply:
Unfortunately no, because then you'd have to perform this check for every possible IP that your host system has. E.g. I have systemd-resolved running on my system, but that one is only reachable via `127.0.0.53` on port 53 udp and tcp. That unfortunately means that `LOCALHOST.socket("udp://53").is_listening` is `False`, however port 53 is still occupied and I cannot bind to it :-/. Thus just getting all open sockets appeared simpler to me than to iterate over all possible ip addresses instead.